pyensign/exceptions.py

```diff
@@ -1,3 +1,93 @@
+import inspect
+from grpc import RpcError
+from functools import wraps
+
+##########################################################################
+## Decorators
+##########################################################################
+
+_except = (RpcError, AttributeError)
+
+
+def wrap_rpc(fn):
+    @wraps(fn)
+    def wrap(*args, **kwargs):
+        try:
+            return fn(*args, **kwargs)
+        except _except as e:
+            handle_client_error(e)
+
+    return wrap
+
+
+def wrap_async_rpc(coro):
+    @wraps(coro)
+    async def wrap(*args, **kwargs):
+        try:
+            return await coro(*args, **kwargs)
+        except _except as e:
+            handle_client_error(e)
+
+    return wrap
+
+
+def wrap_generator(fn):
+    @wraps(fn)
+    def wrap(*args, **kwargs):
+        try:
+            for rep in fn(*args, **kwargs):
+                yield rep
+        except _except as e:
+            handle_client_error(e)
+
+    return wrap
+
+
+def wrap_async_generator(coro):
+    @wraps(coro)
+    async def wrap(*args, **kwargs):
+        try:
+            async for rep in coro(*args, **kwargs):
+                yield rep
+        except _except as e:
+            handle_client_error(e)
+
+    return wrap
+
+
+def catch_rpc_error(coro):
+    """
+    Decorate coroutines to catch and handle gRPC client errors.
+    """
+
+    if inspect.isgeneratorfunction(coro):
+        return wrap_generator(coro)
+    elif inspect.isasyncgenfunction(coro):
+        return wrap_async_generator(coro)
+    elif inspect.iscoroutinefunction(coro):
+        return wrap_async_rpc(coro)
+    else:
+        return wrap_rpc(coro)
+
+
+##########################################################################
+## Error Handling
+##########################################################################
+
+
+def handle_client_error(e):
+    """
+    Provides error handling for errors raised by the gRPC client.
+    """
+
+    if isinstance(e, RpcError):
+        raise EnsignRPCError(e)
+    elif isinstance(e, AttributeError):
+        raise EnsignAttributeError(e)
+    else:
+        raise PyEnsignError(e)
```

Comment:
I'm a little concerned about converting all errors to `PyEnsignErrors` - but I suppose they will all originate from this library (e.g. users won't use `handle_client_error` or `catch_rpc_error` right? If they did; then we could bury other exceptions by accident. 

Perhaps we want to do `_handle_client_error` and `_catch_rpc_error` etc. to indicate to our users that these functions are for internal use only?

Reply:
Yeah they aren't really intended for outside users so I'll mark them with the underscore.

Reply:
Perfect!